### ISSUE TITLE
Better Location button UX

### ISSUE THREAD

**Issue by biodranik** [organicmaps/organicmaps#3450]
Current issues/status for Android (iOS situation is a bit better and different):
1. Some users can't detect their position at all and OM gets one-star reviews and support emails.
2. Some users can't detect precise GPS position, but only a rough position using Internet/WiFi (AGPS), and OM gets one-star reviews and support emails. Both these issues are often caused by timeout interruption of the GPS search, when users are inside buildings or cars, or by screen lock after a timeout that stops position detection.
3. Annoying dialog "continue detecting location" is displayed if the position was not found after some timeout.
4. Many users expect that OM will automatically search for their current position on each app activation like it is done in other popular apps.
5. Some users disable GPS for various reasons and don't want to see an annoying dialog "Please enable location in settings" on each app activation.
6. OM on Android stops any GPS/location activity when goes into the background and starts/continues location detection only on the app reactivation.
7. An active foreground notification service is planned for proper background track recording and background voice navigation/guidance.

Proposed changes for location button states/images:

1. OFF state. Icon: a gray crosshair crossed with a red X. GPS is off, and the current position is unknown. Tap switches the icon/state to (2) if location services and permissions are enabled, or shows an error dialog otherwise. If the state was (1) when OM was closed, then the state is (1) when OM is open again (no warnings if there is no location permission/services). For all other cases, including the first launch, opening OM starts with (2). If the user changes permissions or disables location services while in this mode, there are no warnings/dialogs about it.
2. Active search for satellites/position. Icon: a moving radar. Automatically switches to (3) when GPS or WiFi/Cellular position is found and centers the map. Tap turns off all location providers (stops location search) and returns to (1). Disabling permission/location services in (2) shows an error dialog only if the app was not recently (<2 minutes ago?) backgrounded (in this case there is some reason for the user to launch OM, background it to disable location, and then return back to OM again). If the position was lost in (3), (4), (5), and then found while in (2), the previous mode (3), (4), or (5) should be restored.
3. Position is found, and the map is centered on it. Icon: the inclined blue arrow if it's an accurate GPS position, some symbol is drawn on top if it's a networking position (ideally, an error radius in meters?). Using different GPS/Net icons helps users to understand that their location is inaccurate and reduce negative comments in stores. Tap enables map rotation mode (4). Map dragging switches to (5). The lost position (tunnel, building) switches to (2) but shows the gray arrow on the map (ideally, a time of how old is the position can be displayed on the gray arrow on the map).
4. Map rotation mode. Icon: a filled blue arrow facing north, on top of a vertical line. Tap switches to (3). Map dragging switches to (5). Lost position (tunnel, building) switches to (2).
5. Position is found but not centered. Icon: a gray crosshair. Tap switches to (3). Lost position (tunnel, building) switches to (2).

On Android, every app activation/launch will start from (2) if previously the position was known or was actively searched, and if the background location service was not active (then it restores previously active (3), (4), or (5) mode).

The screen should not dim/lock in modes (2), (3), and (4), and should dim/auto-lock after a system-defined timeout for all apps in (1) and (5).

@vng @rtsisyk @arnaudvergnet @pastk Please provide feedback, this is one of the most important issues for OM.

**Comment by rtsisyk** [comment]
I think that (1) OFF state with a red X should be useful. (2), (3), (4) we already have. (5) I didn't the idea.
I also would like to see the number of satellites and current GPS precision somewhere (Android only). I use GPSTest app for this purpose, but Organic Maps can display some basic information to make users more patient.

**Comment by biodranik** [comment]
We can display satellite count in the rotating radar icon. Or think about a better representation of the GPS quality (percentage?).

**Comment by arnaudvergnet** [comment]
We could display a networks quality icon under/inside the rotating radar icon. The icon would fill in and turn green as the number of satellites/quality increases, and fill out and turn orange/red as the number and quality decreases. Here are example icons from https://materialdesignicons.com/ (keyword "network"):
![Screenshot 2022-09-19 at 11-54-01 Material Design Icons](https://user-images.githubusercontent.com/80701113/190992941-238452e1-ac7f-456b-9528-bd3564cc753f.png)

Maybe this icon could always be present in states 2 to 5. The tricky part would be to find the right placement so it does not clutter the view.

Not everyone knows how many satellites are needed so a simple icon should be easier to understand. There could still be an option for power-users to display the exact number of satellites if you think it is necessary.

A tip for increasing GPS/compass accuracy could also be shown when the detected accuracy is poor, much like explained in https://github.com/organicmaps/organicmaps/issues/3448.


The proposed changes look good to me. I am one of the people who often use the app without GPS like a regular paper map to save battery, and the warning popup on start is always annoying.

For state 1, maybe the red icon would be too strong and mislead the user into thinking their GPS is broken. I would go for a more subtle question mark inside the GPS crosshair from state 5.

**Comment by biodranik** [comment]
As far as I remember, at least 3 satellites are required for GPS to work. So we can replace the rotating radar icon with some blinking satellites ) But let's make it a separate issue (the icon for "searching" state) and focus on other highlighted moments for this issue.

Questionmark for state (1) is misleading. It may mean "looking for, but still unknown". There should be a clear association with "OFF, and is NOT SEARCHING for the location".

**Comment by biodranik** [comment]
Maybe a **GPS** text inside this icon?

![image](https://user-images.githubusercontent.com/170263/192119777-79d9d2fe-43a5-4447-8a3d-d52466b0dfac.png)

**Comment by biodranik** [comment]
Let's start by implementing the OFF state icon/button state (no position and it is not searched). Many users do not understand that this button should be pressed to start a location search.
Some good icons: https://fonts.google.com/icons?icon.query=location&icon.platform=android

CC @rtsisyk

**Comment by ThomasAUS** [comment]
To me this button things is way too overcomplicated. If some users are not turning the gps on then they cannot be helped anyway.

To me this is more a visibility issue:

Try to put your phone on the dashboard of your car, then drive around and tell me if you can see this rotating rarar symbol when the sun is shining - not a chance. 

If you could change this rotating radar symbol to be on an orange background that it is a clear warning that there is no current position, then the main issue is solved. Once this button is orange or red, then it is clear that the displayed position is not updated anymore - job done, and the blue arrow to be on a green background, then all I wanted to achieve is actually done. The functunality is there, it just cannot be readily seen.

Basically all it needs is more colour - I know this is frowned upon nowadays where everything has to be minimalistic grey on grey with a touch more grey. But this is a navigation system - a brilliant one by the way - but it has to be usable outdoors.

Thanks

**Comment by ThomasAUS** [comment]
The basic problem is: if you drive between high buildings or through a tunnel, then the GPS signal gets lost. When the satellites become available again, the GPS has to re-establish the position. but it still shows the last known position as if this was still the current position.
I had a bit more of a think about this today and there might even be an considerably better solution. If you would change from the blue arrow/dot that shows the current position on the map to a bright red one which shows that GPS position is not current. Then the bright red dot could show the last known position. Everybody instinctively looks at the position dot, so that should be the spot where the status is indicated. Or this red circle with the line through it as shown above. It just has to be bright red, that it can clearly be seen.

Thanks

**Comment by biodranik** [comment]
@ThomasAUS thanks for the details! Indeed, there is an issue when the position is not updated, and it should be clear to the user.
What about the following idea: show an increasing number of seconds since the last GPS update in the location button, using some attractive colors (maybe on top of the rotating radar)? Then users will clearly understand that:
1. There is no current GPS signal
2. How old is the location on the screen
3. Location is actively searched for and will be updated automatically as soon as the user goes out of the tunnel or the building.

Of course, proper implementation of this approach should not display "Continue detecting your location dialog?", or should show it and automatically continue searching for it (consider a driver in the car where touching the screen is complicated).

Alternatively, the current position arrow can be changed, but I think that we don't need to complicate it because we already have the state "No GPS signal, but it is actively searched for satellites and will be acquired soon".

**Comment by ThomasAUS** [comment]
Hello Alexander,

It is absolutely important not to have to confirm a loss of signal -
that is just useless and dangerous.

I would really only do the position arrow colour change and leave the
radar symbol alone for the following reasons:

A time since the last update is useless as it will be potentially long
and some small numbers counting will be hard to see anyway. The radar
symbol is about 8mm diameter on my 7" screen. A number over this will be
tiny and virtually invisible.

The position arrow (and I mean the one in the centre of the screen and
not the arrow that replaces the radar symbol) is the thing that people
naturally look at, so why not just a colour change to red and all is clear.

In Australia we have this thing called "KISS" (keep it simple stupid).
Why turn a mouse into an elephant?

I would seriously hope that you _only do the colour change_ and leave
everything else just as is. Or in other words "ingenious is what is
simple and works".

Sorry to dampen down your enthusiasm, but I think if you use your energy
towards the small road visibility, there will be a lot more gained then
by overcomplicating the GPS issue.

Thanks

    Thomas




Am 07.11.2022 um 19:06 schrieb Alexander Borsuk:

**Comment by biodranik** [comment]
I have updated the issue. Let's focus on the final goal for it.

**Comment by biodranik** [comment]
The goal is better UX and it is (partially) described in the ticket description. Let's leave this one as a meta ticket and create new issues for the related subtasks.

It should be very clear for users, what's going on with the location at the moment. Is the signal good? Is location precise or not? What should a user do to get it?

**Comment by ThomasAUS** [comment]
Hello,

Please make sure to make it "visually" clear what the status is. Grey on
grey is not visible with the sun shining onto a display. Red is visible.

Thanks

    Thomas

Am 08.01.2023 um 03:07 schrieb Alexander Borsuk:

**Comment by pastk** [comment]
> The basic problem is: if you drive between high buildings or through a tunnel, then the GPS signal gets lost. When the satellites become available again, the GPS has to re-establish the position. but it still shows the last known position as if this was still the current position. 

When you enter a tunnel the current location stops updating for you, right? I mean the position arrows doesn't move anymore. Isn't it very noticeable? Also when signal is lost the location button should start spinning, which very noticeable also.

I think the biggest issue here is not the color, but a long delay it takes to change from the "position is current" state to the "searching for the position". IMO it won't help much if the blue arrow will change to red after 10-30 seconds of signal being lost (its a long time, esp. if one is driving)..

**Comment by ThomasAUS** [comment]
Hi Konstantin,

I totally agree that no delay is best, but Alexander thought it would be
important for Cell Tower navigation to leave a small time in there.

The example with the tunnel is EXACTLY the problem. It is possible to
see the arrow stopping to move if you look at it all the time. But who
is driving then if the diver has to watch the arrow all the time.
I am old-fashioned and prefer to look at the road and avoid all the
other drivers which always watch their arrows move. I just have a quick
glance at the screen to see where I am at that is it at the most.

The spinning light blue thing is invisible on a bright day as well as
the blue to grey change. Can somebody please please please put a tablet
screen to the dashboard of a car, drive alone in heavy traffic on a
bright sunny day (preferable with the sun shining right on the screen).
Then I'd like to make a guess: You will either not see anything on the
screen or you will have an accident by watching the screen continuously.
Or you have to do it like me, stop, shade the screen with a book to see
what is going on and then continue (great on a busy inner-city road)

I honestly do not understand the problem with this: Signal lost, red
arrow immediately, once re-established it goes back blue, Job done. And
it does not even need any software changes.




Am 27.03.2023 um 19:26 schrieb Konstantin Pastbin:

**Comment by pastk** [comment]
> I honestly do not understand the problem with this: Signal lost, red arrow immediately, once re-established it goes back blue, Job done. And it does not even need any software changes.

As I see it the problem is that it won't be possible to have a "Signal lost, red arrow immediately," behavior unless we somehow get rid of that 10-30 seconds delay...

As for a more contrast arrow color - you also mentioned that white roads are not very visible for you on the sun too. We do get such feedback, but very rarely, which suggests it could depend on a particular device screen. And it seems that for the most of the users the current color scheme is fine contrast-wise. I think the best solution would be to have an optional high-contrast color scheme, but it'll take quite some time to implement...

**Comment by ThomasAUS** [comment]
Hi Konstantin,

An optional high-contrast color scheme is what this app really needs.
I've recommended this app to basically all my relatives and friends
(starting in the Maps.me times), and they all say it is a great thing if
it only would show more contrast. Most are not using it any more, for
mostly these three reasons:

No apparent GPS status shown (Thomas, how do I know when the thing shows
the position?)
The white invisible lines on beige background (Thomas, I don't see the
minor roads, can you change this for me?)
The fact that Freeways/Autobahn have the same colour as main roads and
therefore cannot be seen. (Thomas, I always have to look into the paper
map to find the Freeways/Autobahn.)

But let's just leave it at that. I've mentioned the problem with the
arrow colour and made a suggestion of a solution. Got a reply from
Alexander within an amazing 30 minutes with an implemented working
solution (red arrow after 10 seconds). Ever since this easy and working
solution gets ripped apart for mere design reasons by everyone instead
of just going ahead with it and getting this implemented.

Basically this is just a waste of everyones time and worse than big
politics.

I've really tried to push for Alexanders solution but this will never
happen as Grey is just too important in the 21st century and design is
so much more important than functionality.

If you would do me a really big favour. Send me a link to a beta version
with Alexanders modification in it.

Let me finish with a quote: You don't find out if something works if you
think about it, but if you try it out.

Thomas

PS. So just get out of the office, jump in a car, put the screen on the
dash with the sun reflecting off it and then you tell me that you can
see the difference between grey and blue after looking at it for a short
moment.


Am 27.03.2023 um 21:12 schrieb Konstantin Pastbin:

**Comment by biodranik** [comment]
@ThomasAUS please don't worry, our strategy is to always produce the best solution for most of the users. If the red color is not the best, we'll find another one that will be the best. If you read this issue's description, the issue is more complex than just the "loosing signal" situation.

**Comment by ThomasAUS** [comment]
Hello Konstantin,

I am definitely one of the "niche case of people driving but not using a
navigation mode". I love a paper map, but this is just a super detailed
moving map display which almost always shows me where I am.

So basically the only reference mark I have on the map is the arrow.

That's why I would go to one second delay (probably easier to implement
than 0 seconds) and do a visible colour change (a very light shade of
grey instead of red would even be OK so long as there is a more
significant difference). Maybe light grey would be an acceptable
compromise as it does not "scream" and the gods of grey are appeased.


So I would consider exactly this "niche case" your most important
selling point as this makes Organic Maps so much better than the rest.
Organic Maps displays the map properly (except the white roads :-)
versus with all the other navigation apps, you never have a clue where
you really are or in which direction you are going.

Thank You

   Thomas

Am 29.03.2023 um 19:05 schrieb Konstantin Pastbin:

**Comment by biodranik** [comment]
@pastk @ThomasAUS we plan to introduce #2358 Driving View mode/style for the map. If a user selected it, then obviously we can understand the driving use-case and do a lot of customizations there.

**Comment by oleg-pah** [comment]
Related to #6390

**Comment by biodranik** [comment]
The circle is already drawn in Organic Maps.

**Comment by avidseeker** [comment]
As for signal strength icons, [FitoTrack](https://codeberg.org/jannis/FitoTrack) implementation can be helpful.

**Comment by Jackalaq** [comment]
I'm interested in this button improving the UX so I'll post, just adding that I would love to be able to switch from any state to State 1 (i.e. Turn the location off).
A long press would seem the right solution.
One use case is battery saving when trekking in poor weather and not needing location updates whilst device is in my pocket, but easy use with device in lockscreen mode.

Thanks for considering.
Also requested by others at #4047  such as @thany

**Comment by biodranik** [comment]
@Jackalaq turning off location updates in the system control panel, draggable from the top, can be a more robust/battery efficient solution.
